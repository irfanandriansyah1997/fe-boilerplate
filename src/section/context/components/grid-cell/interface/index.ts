import { MouseEventHandler } from 'react';

/**
 * Grid Cell Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.31
 */
export interface IGridCellProps {
  cell: number;
  onClickCell: MouseEventHandler<HTMLButtonElement>;
}
