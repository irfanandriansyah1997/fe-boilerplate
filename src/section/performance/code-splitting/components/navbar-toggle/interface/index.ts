/**
 * Navbar Toggle Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.14
 */
export interface INavbarToggleProps {
  active: boolean;
  onToggle(active: boolean): void;
}
