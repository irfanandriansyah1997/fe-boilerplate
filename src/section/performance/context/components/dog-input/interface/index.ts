import { ChangeEventHandler } from 'react';

/**
 * Dog Input Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.31
 */
export interface IDogInputProps {
  dogName: string;
  onChangeDogName: ChangeEventHandler<HTMLInputElement>;
}
