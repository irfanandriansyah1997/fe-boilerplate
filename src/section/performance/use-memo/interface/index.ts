/**
 * City Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.28
 */
export interface ICity {
  country: string;
  id: string;
  lat: string;
  lng: string;
  name: string;
}

/**
 * City V2 Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.28
 */
export interface ICityV2 extends ICity {
  isActive?: boolean;
}
