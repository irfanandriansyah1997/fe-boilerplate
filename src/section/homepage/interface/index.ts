import { IMasonryItem } from '../../../components/organisms/masonry-layout/interface';

/**
 * App Menu List Item Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.11.17
 */
export interface IAppMenuListItem {
  dataTestID: string;
  item: IMasonryItem[];
  label: string;
}

/**
 * App Menu List Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.11.17
 */
export interface IAppMenuList {
  section: IAppMenuListItem[];
}
