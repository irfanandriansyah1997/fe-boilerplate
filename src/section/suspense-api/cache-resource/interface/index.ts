import { IResourcePayload } from '../../../../interface/general';
import { IPokemonDetail } from '../../../../interface/pokemon';

/**
 * Pokemon Cache Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.11.21
 */
export type IPokemonCache = (
  pokemonName: string
) => IResourcePayload<IPokemonDetail>;

/**
 * Pokemon Provider Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.11.21
 */
export interface IPokemonProviderProps {
  cacheTime?: number;
}
