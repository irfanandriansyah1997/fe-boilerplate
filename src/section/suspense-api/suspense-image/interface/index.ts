import { IResourcePayload } from '../../../../interface/general';
import { IPokemonDetail } from '../../../../interface/pokemon';

/**
 * Pokemon Cache Payload Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.11.21
 */
export interface IPokemonCachePayload {
  data: IResourcePayload<IPokemonDetail>;
  image: IResourcePayload<string>;
}

/**
 * Pokemon Cache Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.11.21
 */
export type IPokemonCache = (pokemonName: string) => IPokemonCachePayload;

/**
 * Pokemon Provider Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.11.21
 */
export interface IPokemonProviderProps {
  cacheTime?: number;
}
