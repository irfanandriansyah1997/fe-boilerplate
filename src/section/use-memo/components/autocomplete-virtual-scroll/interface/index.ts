import { LegacyRef, ReactNode, ReactText } from 'react';
import { VirtualItem } from 'react-virtual/types';

import { ICity, ICityV2 } from '../../../interface';

/**
 * Autocomplete V2 Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.29
 */
export interface IAutocompleteV2Props {
  children?: ReactNode;
  items: ICityV2[];
  keyword: string;
  listRef: LegacyRef<HTMLDivElement>;
  resetSelectedItem(): void;
  setKeyword(param: string): void;
  setSelectedItem(param: ICityV2): void;
  totalItems: number;
  virtualItems: VirtualItem[];
}

/**
 * Autocomplete V2 Item Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.30
 */
export interface IAutocompleteV2ItemProps extends ICity {
  active?: boolean;
  onClick?: (param: ICityV2) => void;
  style: IAutocompleteV2ItemStyling;
}

/**
 * Autocomplete V2 Item Styling Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.30
 */
export interface IAutocompleteV2ItemStyling {
  size: ReactText;
  start: number;
}
