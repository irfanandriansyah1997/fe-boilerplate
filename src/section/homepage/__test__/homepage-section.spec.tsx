import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import { IAppMenuList } from '../interface';
import Homepage from '..';

jest.mock(`../helper/index`, () => ({
  generateAppMenu: jest.fn().mockImplementation(
    (): IAppMenuList => ({
      section: [
        {
          dataTestID: `homepage-section`,
          item: [
            {
              item: {
                menu: [
                  {
                    text: `Wrap Use Memo`,
                    to: `/context/part-1`
                  },
                  {
                    text: `Separate Context Value`,
                    to: `/context/part-2`
                  },
                  {
                    text: `Colocate State`,
                    to: `/context/part-3`
                  }
                ],
                subtitle: `Sample Subtitle Course`,
                title: `Title Course`
              },
              key: `key 1`
            }
          ],
          label: `Performance`
        }
      ]
    })
  )
}));

describe(`Testing Homepage Section`, () => {
  it(`Snapshot Testing`, () => {
    const { container } = render(
      <MemoryRouter>
        <Homepage />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });

  it(`Should be rendered title and description on homepage`, () => {
    const { getByTestId } = render(
      <MemoryRouter>
        <Homepage />
      </MemoryRouter>
    );

    expect(getByTestId(/homepage-title/)).toHaveTextContent(
      `React.js cheatsheet`
    );
    expect(getByTestId(/homepage-description/)).toHaveTextContent(
      `React is a JavaScript library for building user interfaces. This guide targets React v15 to v16.`
    );
  });

  it(`Should be rendered one card on homepage`, () => {
    const { getAllByTestId, queryAllByTestId, queryByText } = render(
      <MemoryRouter>
        <Homepage />
      </MemoryRouter>
    );

    expect(getAllByTestId(/^homepage-card$/i)).toHaveLength(1);
    expect(queryByText(/^Title Course$/i)).toBeInTheDocument();
    expect(queryByText(/^Sample Subtitle Course$/i)).toBeInTheDocument();

    expect(queryAllByTestId(/homepage-card-link/)).toHaveLength(3);
    const [firstElement, secondElement, thirdElement] = queryAllByTestId(
      /homepage-card-link/
    );

    /**
     * Check Attribute First Item
     */
    expect(firstElement).toHaveTextContent(`sticky_note_2Wrap Use Memo`);
    expect(firstElement).toHaveAttribute(`href`, `/context/part-1`);

    /**
     * Check Attribute Second Item
     */
    expect(secondElement).toHaveTextContent(
      `sticky_note_2Separate Context Value`
    );
    expect(secondElement).toHaveAttribute(`href`, `/context/part-2`);

    /**
     * Check Attribute Third Item
     */
    expect(thirdElement).toHaveTextContent(`sticky_note_2Colocate State`);
    expect(thirdElement).toHaveAttribute(`href`, `/context/part-3`);
  });
});
