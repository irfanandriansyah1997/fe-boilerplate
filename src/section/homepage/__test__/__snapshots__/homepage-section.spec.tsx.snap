// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Homepage Section Snapshot Testing 1`] = `
<div>
  <div
    class="homepage"
  >
    <div
      class="homepage__heading"
    >
      <h1
        data-testid="homepage-title"
      >
        React.js cheatsheet
      </h1>
      <p
        data-testid="homepage-description"
      >
        React is a JavaScript library for building user interfaces. This guide targets React v15 to v16.
      </p>
    </div>
    <div
      class="homepage__section"
      data-testid="homepage-section"
    >
      <h2>
        Performance
      </h2>
      <div
        class="o-masonry"
        data-testid="masonry-container"
        style="grid-template-columns: repeat(2, 1fr);"
      >
        <div
          data-testid="masonry-column"
        >
          <div
            class="homepage-card"
            data-testid="homepage-card"
          >
            <h6>
              Title Course
            </h6>
            <div
              class="homepage-card__container"
            >
              <p>
                Sample Subtitle Course
              </p>
              <ul>
                <li>
                  <a
                    data-testid="homepage-card-link"
                    href="/context/part-1"
                  >
                    <span
                      class="material-icons"
                    >
                      sticky_note_2
                    </span>
                    Wrap Use Memo
                  </a>
                </li>
                <li>
                  <a
                    data-testid="homepage-card-link"
                    href="/context/part-2"
                  >
                    <span
                      class="material-icons"
                    >
                      sticky_note_2
                    </span>
                    Separate Context Value
                  </a>
                </li>
                <li>
                  <a
                    data-testid="homepage-card-link"
                    href="/context/part-3"
                  >
                    <span
                      class="material-icons"
                    >
                      sticky_note_2
                    </span>
                    Colocate State
                  </a>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
